Porting System.Net.Security

CIPop opened the pull request:
Missing issues for TODO comments. Will create the tracking issue and update the code, author (dotnetbot) and reference the tracking issue.

@davidsh @bartonjs @pgavlin @stephentoub @SidharthNabar  PTAL.

davidsh commented on src/System.Net.Security/src/System/IO/StreamAsyncHelper.cs:
```diff
@@ -0,0 +1,193 @@
+{
+    /// <summary>Provides support for implementing asynchronous operations on Streams.</summary>
+    internal sealed class StreamAsyncHelper
+    {
```
Why is the `StreamAsyncHelper` class buried in this dll?  Why not put into src/Common somewhere? Then it could be reused by others.

CIPop replied:
Actually this was in Common\IO\StreamAsyncHelper and got removed by @Maxwe11 in 2732ffa58073e287fc68f5a1623cbf4c0700f44c just before I've finished this work.

Do you think I should move this back?

CIPop replied:
Moved to common.

davidsh commented on src/System.Net.Security/src/System/Net/SecureProtocols/_FixedSizeReader.cs:
```diff
@@ -0,0 +1,144 @@
+    // Reads a fixed size packet from a stream, can disover EOF as 0 bytes read from stream.
+    //
+    internal class FixedSizeReader
+    {
```
Why is this file called "_FixedSizeReader.cs"?  I.e. why use the underscore character at all?  I understand this was ported from Desktop.  Is it using an underscore because it is an internal class? 

@stephentoub Most of current code in CoreFx doesn't appear to use files with underscores in them. I wonder if we have any guidance of naming like this?

CIPop replied:
I agree. 41 files have this naming in our staging branch across multiple namespaces. Tracking with #3116

stephentoub replied:
> Most of current code in CoreFx doesn't appear to use files with underscores in them. I wonder if we have any guidance of naming like this?

We've tried to remove underscores from file names in cases like this.  It's possible some slipped through or we just didn't finish cleaning everything up.

davidsh commented on src/System.Net.Security/src/System/Net/SecureProtocols/_SecuritySafeHandles.Windows.cs:
```diff
@@ -0,0 +1,1339 @@
+            _handle = new Interop.Secur32.SSPIHandle();
+        }
+
+#if TRAVE
```
"TRAVE"....I kinow this is being ported.  But it would be good to have another name that we standardize across CoreFx for this.  "TRAVE"  == "TRAce VErbose".

CIPop replied:
Has been discussed before. https://github.com/dotnet/corefx/search?q=TRAVE&type=Issues&utf8=%E2%9C%93

If you believe we should remove or rename it, please open a tracking issue.

stephentoub replied:
I do think a different name would be good, and to standardize that across corefx.  My biggest issue with "TRAVE" is that it really looks like a typo, especially since v is next to c on a qwerty keyboard.

CIPop replied:
Discussion opened at #3163. Please add suggestions, thoughts there. After we decided on something, we'll need to change the entire repo and ensure that the rule is present in our coding guidelines.

davidsh commented on src/System.Net.Security/src/project.json:
```diff
@@ -0,0 +1,10 @@
+  "dependencies": {
+    "System.Diagnostics.Contracts": "4.0.0-*",
+    "System.Threading": "4.0.0-*",
+    "System.Net.Primitives": "4.0.10-beta-*"
```
cc: @ericstj 

I'm not sure we should still be basing on "-beta-" in the version fields anymore.

CIPop replied:
Replacing with 4.0.10-*" will break the build at this time although not because of Primitives types missing, but X509Certificate types. 

Since the dependency tree is different, I'd like to wait for @ericstj's answer on this before any further changes.

My current plan is to review all references once we enable individual packages for System.Net contracts.

davidsh commented on src/Common/src/System/Net/DebugCriticalHandleMinusOneIsInvalid.cs:
```diff
@@ -0,0 +1,45 @@
+    // As a general rule normal code path should always destroy handles explicitly
+    //
+    internal abstract class DebugCriticalHandleMinusOneIsInvalid : CriticalHandleMinusOneIsInvalid
+    {
```
Why create a separate class for this `DebugCriticialhandleMinusOneIsInvalid` ?  Why not just put this if DEBUG stuff into the main `CriticalHandleMinusOneIsInvalid` class?  Right now, this is buried in System.Net files and it could be useful for everyone else.  Plus separate files are harder to track and we have to remember to include them separately in the CSPROJ files for DEBUG vs. RETAIL.

CIPop replied:
The only reason is that most of these classes do logging that is specific to System.Net and so can't be shared unless the code takes dependency on our GlobalLog ETW tracer.

Maxwe11 commented on src/Common/src/System/Net/_LazyAsyncResult.cs:
```diff
@@ -399,7 +399,7 @@ protected virtual void Complete(IntPtr userToken)
                         Task.Factory.StartNew(
-                            s => WorkerThreadComplete(s), 
+                            s => WorkerThreadComplete(s),
                             null,
```
Could we pass `this` here and avoid its capturing?

CIPop replied:
Converted the callback to static and passing `this` instead of `null`.

Maxwe11 commented on src/Common/tests/System.Net/TaskAPMExtensions.cs:
```diff
@@ -0,0 +1,83 @@
+            {
+                if (callback != null)
+                {
+                    task.ContinueWith(delegate { callback(task); },
```
Most of the code in corefx uses lambdas. Why anonymous method used here?

CIPop replied:
Instead of changing this, I've added a TODO to completely remove this and rewrite the tests using TPL: #3144

This is a temporary workaround to avoid rewriting and stabilizing all tests using TPL instead of APM.

davidsh commented:
LGTM

pgavlin commented on src/Common/src/System/IO/StreamAsyncHelper.cs:
```diff
@@ -0,0 +1,193 @@
+
+        internal IAsyncResult BeginRead(byte[] buffer, int offset, int count, AsyncCallback callback, object state)
+        {
+            if (!_stream.CanRead) throw __Error.GetReadNotSupported();
```
Have these files been formatted?

pgavlin replied:
To expand a bit, I think that this line should be

```
if (!_stream.CanRead)
{
    throw __Error.GetReadNotSupported();
}
```

There are a number of other instances as well.

CIPop replied:
Good catch: this file got deleted during work I did for this PR. This is just bringing it back to the same location.

It appears that the file never got formatted for some reason. I'll run the tools again.

CIPop replied:
Not handled by CodeFormatter. Will manually change.

pgavlin commented:
LGTM aside from some code style issues.

stephentoub commented on src/Common/src/Interop/Windows/secur32/Interop.SSPI.cs:
```diff
@@ -0,0 +1,509 @@
+                aIssuers = handle;
+                fixed (byte* voidPtr = nativeBuffer)
+                {
+                    // if this breaks on 64 bit, do the sizeof(IntPtr) trick
```
Is this comment still relevant?

CIPop replied:
The comment is from the original code. I'd keep it for now since it uncovers an issue in marshaling the struct from a native buffer: There is an assumption made that padding won't be added on any any architecture. 
Changed the comment to reflect this and linked to #3114.

stephentoub commented on src/Common/src/System/Net/_LazyAsyncResult.cs:
```diff
@@ -430,15 +430,18 @@ protected virtual void Complete(IntPtr userToken)
 
         // Only called in the above method
-        private void WorkerThreadComplete(object state)
+        private static void WorkerThreadComplete(object state)
```
This change is fine, but if you wanted to leave this method as it was, you could instead have just made the call site above use:
```C#
s => ((LazyAsyncResult)s).WorkerThreadComplete(), this
```

CIPop replied:
Thanks Stephen. I'll leave the callback as static for now.

stephentoub commented on src/System.Net.Security/src/System/Net/SecureProtocols/AuthenticatedStream.cs:
```diff
@@ -0,0 +1,84 @@
+
+namespace System.Net.Security
+{
+    // A public contact for a base abstract authenticated stream.
```
"public contract"?

CIPop replied:
Good catch. Corrected.

stephentoub commented on src/System.Net.Security/src/System/Net/SecureProtocols/AuthenticatedStream.cs:
```diff
@@ -0,0 +1,84 @@
+    public abstract class AuthenticatedStream : Stream
+    {
+        private Stream _InnerStream;
+        private bool _LeaveStreamOpen;
```
Nit: fields should be camel cased, e.g. ```_leaveStreamOpen``` instead of ```_LeaveStreamOpen```.

CIPop replied:
I thought the tooling will automatically do that. I'll run it again. (@pgavlin found another place where style wasn't updated.)

I'll open bugs if these files are not changed and make the corrections manually

CIPop replied:
Will manually change. Tracking with https://github.com/dotnet/codeformatter/issues/177.

stephentoub commented on src/System.Net.Security/src/System/Net/SecureProtocols/AuthenticatedStream.cs:
```diff
@@ -0,0 +1,84 @@
+                        }
+                        else
+                        {
+                            _InnerStream.Dispose();
```
We're not flushing in this case... I guess the assumption is that Dispose on the innerStream will flush itself if it needs to?

CIPop replied:
This is the original code. I'll keep the behavior for app-compat purposes.

stephentoub commented on src/System.Net.Security/src/System/Net/SecureProtocols/SslStream.cs:
```diff
@@ -0,0 +1,455 @@
+        {
+        }
+
+        public SslStream(Stream innerStream, bool leaveInnerStreamOpen)
```
Nit: we seem to be inconsistent about blank lines between ctors

CIPop replied:
Unfortunately tooling didn't fix this... Looks like I missed a few places.

stephentoub commented on src/System.Net.Security/src/System/Net/SecureProtocols/SslStream.cs:
```diff
@@ -0,0 +1,455 @@
+            _SslState = new SslState(innerStream, _userCertValidationCallbackWrapper, _userCertSelectionCallbackWrapper, encryptionPolicy);
+        }
+
+        private bool UserCertValidtaionCallbackWrapper(string hostName, X509Certificate2 certificate, X509Chain chain, SslPolicyErrors sslPolicyErrors)
```
Nit: Validtaion => Validation

CIPop replied:
Fixed.

stephentoub commented on src/System.Net.Security/src/System/Net/SecureProtocols/SslStream.cs:
```diff
@@ -0,0 +1,455 @@
+
+        private SslProtocols DefaultProtocols()
+        {
+            SslProtocols protocols = SslProtocols.Tls12 | SslProtocols.Tls11 | SslProtocols.Tls;
```
const?

CIPop replied:
Will change to const.

@SidharthNabar :  We should also make this function public and recommend that apps use it instead of hardcoding the default protocol list.

stephentoub commented on src/System.Net.Security/src/System/Net/SecureProtocols/SslStream.cs:
```diff
@@ -0,0 +1,455 @@
+        public override void Write(byte[] buffer, int offset, int count)
+        {
+            _SslState.SecureStream.Write(buffer, offset, count);
+        }
```
Is it possible the inner stream overrides ReadAsync/WriteAsync/FlushAsync?  If so, this stream should ideally override those as well and delegate to the inner counterparts.

CIPop replied:
Tested a few things but I still don't understand the comment.

Implementing *Async overrides isn't going to help with the inner stream which is wrapped in `_SslState.SecureStream`. Since that's not a `Stream` (`_SslStream` doesn't derive from anything) and since it doesn't implement any TPL, the overrides will only do what the base (`Stream`) already does to adapt Write/Read/Flush to Async.

SslState is dependent on APM it at this time. Since CoreFX Stream doesn't have that anymore, I (temporarily) rerouted the async calls to `SslState._InnerStreamAPM` via `StreamAsyncHelper`. The helper should already call the proper ReadAsync/WriteAsync/FlushAsync APIs. I'm tracking proper TPL implementations and removing the helper with #3114.

stephentoub replied:
I may just not understand what these types are.  SecureStream returns a type derived from Stream, right?  If no, then ignore my comment. If yes, then I was asking if that Stream overrides ReadAsync/WriteAsync/FlushAsync, and if it does, then there's value here in also overriding ReadAsync/WriteAsync/FlushAsync... if it doesn't, then there's nothing to be done.

CIPop replied:
@stephentoub Thanks for confirming! The `_SslState.SecureStream` object is _not_ a `Stream`. 
It's a `_SslStream` which doesn't derive from anything. (The name is confusing for many reasons.)

```
    internal class _SslStream
    { [...]

```

stephentoub commented on src/System.Net.Security/src/System/Net/SecureProtocols/SslStreamContext.cs:
```diff
@@ -0,0 +1,24 @@
+            return _sslStream.GetChannelBinding(kind);
+        }
+
+        private SslStream _sslStream;
```
reaodnly?

CIPop replied:
Fixed.

stephentoub commented on src/System.Net.Security/src/System/Net/SecureProtocols/_SecuritySafeHandles.Windows.cs:
```diff
@@ -0,0 +1,1340 @@
+            {
+                if (b)
+                {
+                    phContext.DangerousRelease();
```
I think this is dead code.   b won't be true here.

CIPop replied:
Good catch: I'll scan for all other places where AddRef/Release are used.

In a previous discussion with @davidsh we've found that since in CoreFX ThreadAbort can't be caught, the entire pattern can be removed.

stephentoub commented on src/System.Net.Security/src/System/Net/SecureProtocols/_SecuritySafeHandles.Windows.cs:
```diff
@@ -0,0 +1,1340 @@
+        //This method should never be called for this type
+        public new IntPtr DangerousGetHandle()
+        {
+            throw new InvalidOperationException();
```
Maybe Debug.Fail / Environment.FailFast?  Code could erroneously catch this exception making it more difficult to spot that there's a problem.

CIPop replied:
Makes sense. I'll switch to Debug.Fail since the code is already under a `DEBUG` ifdef.

stephentoub commented on src/System.Net.Security/src/System/Net/SecureProtocols/_SecuritySafeHandles.Windows.cs:
```diff
@@ -0,0 +1,1340 @@
+    {
+        public SafeFreeCredential_SECURITY() : base() { }
+
+        override protected bool ReleaseHandle()
```
Nit: we're a bit inconsistent on the order of the visibility and override

davidsh replied:
I always thought it was visibility override.  In sample code I have always seen "protected virtual" or "protected overrirde" and never this: "override protected".

CIPop replied:
Yes. Again, the automatic tooling didn't detect/change this. I'll search for other places.

CIPop replied:
Tracking CodeFormatter issue with https://github.com/dotnet/codeformatter/issues/177.

Maxwe11 replied:
FYI you can fix issues like this directly in VS via https://github.com/DotNetAnalyzers/StyleCopAnalyzers it also implements code fixes.

stephentoub commented on src/System.Net.Security/src/System/Net/SecureProtocols/_SecuritySafeHandles.Windows.cs:
```diff
@@ -0,0 +1,1340 @@
+                if (b2)
+                {
+                    outContext.DangerousRelease();
+                    b2 = false;
```
Another case of dead code I believe.

stephentoub commented on src/System.Net.Security/src/System/Net/SecureProtocols/_SslState.cs:
```diff
@@ -0,0 +1,1745 @@
+                if (obj is LazyAsyncResult)
+                {
+                    // Sync handshake is waiting on other thread.
+                    ((LazyAsyncResult)obj).InvokeCallback();
```
It's ok that the callback is being invoked while holding the lock?  No opportunities for deadlock as a result?

CIPop replied:
This requires further investigation and testing. Adding to #3114.

stephentoub commented on src/System.Net.Security/src/System/Net/_SslSessionsCache.cs:
```diff
@@ -0,0 +1,217 @@
+                {
+                    return true;
+                }
+                if ((object)sslCredKey1 == null || (object)sslCredKey2 == null)
```
SslCredKey is a struct.  Seems like the first if will end up boxing the keys, such that they won't have reference equality, and the second if will box them and then never find them null.  So allocation that ends up being a nop?

stephentoub commented on src/System.Net.Security/tests/FunctionalTests/project.json:
```diff
@@ -0,0 +1,16 @@
+
+    "xunit": "2.0.0-beta5-build2785",
+    "xunit.abstractions.netcore": "1.0.0-prerelease",
+    "xunit.assert": "2.0.0-beta5-build2785",
```
Are these the right xunit versions to be targeting?

CIPop replied:
Copied from the wrong place. Replaced with 2.1.0 beta3 references.

stephentoub commented:
Thanks, @CIPop.  Left a bunch of small comments, but nothing to prevent this from being merged. LGTM.

CIPop commented:
@davidsh, @pgavlin  @stephentoub Thanks for the thorough review. I've made changes to address most comments and opened tracking issues for others.

PTAL (especially for the SafeHandle transform).

(Will squash and change author for the last version.)

bartonjs commented on src/System.Net.Security/src/System/Net/SecureProtocols/SslStream.cs:
```diff
@@ -0,0 +1,454 @@
+                return (sslPolicyErrors == SslPolicyErrors.None);
+            }
+            else
+                return _userCertificateValidationCallback(this, certificate, chain, sslPolicyErrors);
```
In a comment in the middle of the issue linked to for our bracing rules it says that we like all the statements of the same if group to have braces-or-not; so this return would be braced to match the if being braced.

I don't know how well we enforce that; but it did catch me as a little off while reading it.

CIPop replied:
Fixing all manually. We prefer all `if` statements to have braces.

bartonjs commented on src/System.Net.Security/src/System/Net/SecureProtocols/SslStream.cs:
```diff
@@ -0,0 +1,454 @@
+                return _userCertificateValidationCallback(this, certificate, chain, sslPolicyErrors);
+        }
+
+        private X509Certificate userCertSelectionCallbackWrapper(string targetHost, X509CertificateCollection localCertificates, X509Certificate remoteCertificate, string[] acceptableIssuers)
```
Is this method intentionally camelCased (instead of PascalCased)?

CIPop replied:
Unknown: brought in from original code. The new coding style requires PascalCased: changing an searching for others.

bartonjs commented on src/System.Net.Security/src/System/Net/SecureProtocols/SslStream.cs:
```diff
@@ -0,0 +1,454 @@
+        //
+        public virtual void AuthenticateAsClient(string targetHost)
+        {
+            AuthenticateAsClient(targetHost, new X509CertificateCollection(), DefaultProtocols, false);
```
I looked ahead to see if `null` was allowed for `clientCertificates` and it's eventually allocated as `new X509CertificateCollection` if `null` is passed in.  Personally I'd pass `null` here and let it get initialized later; it makes it clear that `null` is legal to the overload as well as save a mild amount of GC work in the event `targetHost` (etc) is invalid and throws.

bartonjs commented on src/System.Net.Security/src/System/Net/SecureProtocols/SslStream.cs:
```diff
@@ -0,0 +1,454 @@
+
+                object chkCertificateOrBytes = _remoteCertificateOrBytes;
+                if (chkCertificateOrBytes != null && chkCertificateOrBytes.GetType() == typeof(byte[]))
+                    return (X509Certificate)(_remoteCertificateOrBytes = new X509Certificate((byte[])chkCertificateOrBytes));
```
I assume the return type is locked to X509Certificate(1) because of legacy constraints, but why only allocate it as an X509Certificate(1) object instead of X509Certificate2?

bartonjs replied:
("Because Desktop does and it would make for bad code where sometimes people cast blindly and other times they do `as`" is a perfectly good reason, I'm just asking the question :smile:)

CIPop replied:
You are correct: this is because the property is in the public contract and it needs to match Desktop to allow PLib development.

bartonjs commented on src/System.Net.Security/src/System/Net/SecureProtocols/_FixedSizeReader.cs:
```diff
@@ -0,0 +1,144 @@
+        private readonly Stream _Transport;
+        private AsyncProtocolRequest _Request;
+        private int _TotalRead;
+
```
All of these fields are _PascalCased instead of _camelCased.  Is that something that CodeFormatter missed, or was it intentional?

CIPop replied:
This is among many things that CodeFormatter is incorrectly changing.

bartonjs commented on src/System.Net.Security/src/System/Net/SecureProtocols/_FixedSizeReader.cs:
```diff
@@ -0,0 +1,144 @@
+
+                if (bytes == 0)
+                {
+                    if (tempCount != count)
```
I don't quite follow this.  It feels like asking for 1024 bytes when 12 remain would do a loop that read 12, then throw.  Is that right? Is that good?  What happened to the last 12 bytes? (Oh, won't somebody think of the 12 bytes?!)

Some comments might be warranted :smile:.

CIPop replied:
Changed the comment with the one in the original Abstract:
```C#
    //
    // The class is a simple wrapper on top of a read stream. It will read the exact number of bytes requested.
    // It will throw if EOF is reached before the expected number of bytes is returned.
    //
    internal class FixedSizeReader
```

From my understanding of the comments and code: this will read the exact requested size or throw. 

If the protocol was expecting 1024 bytes but 12 were returned and then EOF was hit, then this is a protocol error and the class will throw.

Your comment is revealing the need for Unit Tests for this and other classes.

bartonjs commented on src/System.Net.Security/src/System/Net/SecureProtocols/_SecuritySafeHandles.Windows.cs:
```diff
@@ -0,0 +1,1219 @@
+#endif
+        protected SafeFreeContextBuffer() : base(true) { }
+
+        // This must be ONLY called from this file and in the context of a CER
```
Aren't ConstrainedExecutionRegions a thing of the past?

Also, the method doesn't really need the `unsafe` keyword, it uses IntPtr, not void*.

CIPop replied:
Good point: removed `unsafe`. Removed references to CER / Constrained* from all comments.

bartonjs commented on src/System.Net.Security/src/System/Net/SecureProtocols/_SecuritySafeHandles.Windows.cs:
```diff
@@ -0,0 +1,1219 @@
+        internal static int EnumeratePackages(out int pkgnum, out SafeFreeContextBuffer pkgArray)
+        {
+            int res = -1;
+            SafeFreeContextBuffer_SECURITY pkgArray_SECURITY = null;
```
None of the instance methods on this type seem to check for IsInvalid (and/or IsClosed).  Is that expected?

CIPop replied:
Not sure I understand the comment. EnumeratePackages looks like a factory method: it creates a `SafeFreeContextBuffer`. The handle returned can be invalid:
```C#
            SafeFreeContextBuffer_SECURITY pkgArray_SECURITY = null;
            res = Interop.SafeNetHandles_SECURITY.EnumerateSecurityPackagesW(out pkgnum, out pkgArray_SECURITY);
            pkgArray = pkgArray_SECURITY;

            if (res != 0 && pkgArray != null)
            {
                pkgArray.SetHandleAsInvalid();
            }
```

bartonjs replied:
Oops, I missed that the handles were all outbound instead of inbound; I was expecting  an `IsInvalid` before the P/Invoke, but that'd be completely backwards.  My bad.

bartonjs commented on src/System.Net.Security/src/System/Net/SecureProtocols/_SecuritySafeHandles.Windows.cs:
```diff
@@ -0,0 +1,1219 @@
+                                                    string package,
+                                                    Interop.Secur32.CredentialUse intent,
+                                                    out SafeFreeCredentials outCredential
+                                                    )
```
I don't really grok the indentation here.

CIPop replied:
Another miss from CodeFormatter.

bartonjs commented on src/System.Net.Security/src/System/Net/SecureProtocols/_SecuritySafeHandles.Windows.cs:
```diff
@@ -0,0 +1,1219 @@
+    // Implementation of handles that are dependent on DeleteSecurityContext
+    //
+#if DEBUG
+    internal abstract class SafeDeleteContext : DebugSafeHandle
```
I've put several SafeHandle types in one file myself; but they were all just ctor+IsInvalid+ReleaseHandle.  There's a lot of code in these types, maybe it would make sense to split the file up.

The instance methods also make it complicated on the to-be-written PAL, since there's not really a good place to insert interfaces to help with signature alignment.

CIPop replied:
Tracking with #3114.

bartonjs commented on src/System.Net.Security/src/System/Net/SecureProtocols/_SslState.cs:
```diff
@@ -0,0 +1,1745 @@
+        private FixedSizeReader _Reader;
+
+        private int _NestedAuth;
+        private SecureChannel _Context;
```
Lots more _PascalCased fields.

CIPop replied:
Manually fixed. I've opened a bug for this (see above comments).

bartonjs commented on src/System.Net.Security/src/System/Net/SecureProtocols/_SslState.cs:
```diff
@@ -0,0 +1,1745 @@
+            ValidateCreateContext(isServer, targetHost, enabledSslProtocols, serverCertificate, clientCertificates, remoteCertRequired,
+                                   checkCertRevocationStatus, !isServer);
+        }
+        internal void ValidateCreateContext(bool isServer, string targetHost, SslProtocols enabledSslProtocols, X509Certificate serverCertificate, X509CertificateCollection clientCertificates, bool remoteCertRequired, bool checkCertRevocationStatus, bool checkCertName)
```
These couple of methods don't have spacing between them, but the rest do

CIPop replied:
Thanks! Missed these two.

bartonjs commented on src/System.Net.Security/src/System/Net/SecureProtocols/_SslState.cs:
```diff
@@ -0,0 +1,1745 @@
+                if ((proto & SslProtocols.Ssl2) != 0)
+                {
+                    proto |= SslProtocols.Ssl2;
+                }
```
I don't understand what these ifs are doing.

If the SSL2 bit is set, set the SSL2 bit.  Repeat for further protocols.  The comment sounds like it's trying to clear out unknown bits; but that seems like you want a read-from enum variable and a write-to enum variable...

CIPop replied:
I'll fix this.

bartonjs commented on src/System.Net.Security/src/System/Net/SecureProtocols/_SslState.cs:
```diff
@@ -0,0 +1,1745 @@
+
+        private enum Framing
+        {
+            None = 0,
```
Based on usage, wouldn't `Unknown` be better than `None`?

CIPop replied:
Makes sense: internal code and so can be safely changed.

bartonjs commented on src/System.Net.Security/src/System/Net/SecureProtocols/_SslState.cs:
```diff
@@ -0,0 +1,1745 @@
+            if ((bytes[0] & 0x80) == 0)
+                // We have a three-byte header format
+                GlobalLog.Print("WARNING: SslState::DetectFraming() SSL v <=2 HELLO has no high bit set for 3 bytes header, we are broken, received byte = " + bytes[0].ToString("x", NumberFormatInfo.InvariantInfo));
+#endif
```
I don't understand why we'd expect `bytes[0]` to have the high bit set.  The description in the comment blocks above is that it's the MSB, so that being set requires it to be encoding a length larger than 32767, but the comments keep talking about "3".

CIPop replied:
Note that version is `bytes[1] << 8` but that is done after checking `bytes[1] != 3` (For the Ssl3 case.)

The `TRACE_VERBOSE` comment is made for the other case. I'm no expert but the comments above say that the handshake is started with `RECORD_LENGTH_MSB` which is something that should be ignored for all other protocols. 

This is stating that if the high bit is set in the MSB record length the negotiation is a 3 byte header.

http://www.ietf.org/mail-archive/web/tls/current/msg15243.html:

>In SSLv2 the bytes of the record header are:
>0: record length MSB, & 0x7f (or & 0x3f)
>1: record length LSB
><optional padding length>

The current state seems correct although I don't have access to SSLv2 documentation to confirm.

bartonjs commented on src/System.Net.Security/src/System/Net/SecureProtocols/_SslState.cs:
```diff
@@ -0,0 +1,1745 @@
+                version = (bytes[1] << 8) | bytes[2];
+                if (version < 0x300 || version >= 0x500)
+                {
+                    return Framing.Invalid;
```
Is it not possible to have had an SSLv2 packet which starts with `{ 22, 6, 1 }`?

CIPop replied:
I'm not sure but it doesn't seem like that's possible (see above).

bartonjs commented on src/System.Net.Security/src/System/Net/SecureProtocols/_SslState.cs:
```diff
@@ -0,0 +1,1745 @@
+            // When server has replied the framing is already fixed depending on the prior client packet
+            if (!Context.IsServer || _Framing == Framing.Unified)
+            {
+                return Framing.BeforeSSL3;
```
I thought we were eliminating SSLv2 from CoreFX?

CIPop replied:
/cc: @vijaykota, @rajansingh10 

Good point: SSL2 isn't supported by the OS by default but it can be enabled via registry. Removing SSLv2 from this implementation is out of scope for this PR so I'll leave this for a future change. 

Tracking in #3114.

vijaykota replied:
Behavior on Ubuntu is similar. SSLv2 is disabled but libssl can be rebuilt with explicit SSLv2 support

bartonjs commented on src/System.Net.Security/src/System/Net/SecureProtocols/_SslState.cs:
```diff
@@ -0,0 +1,1745 @@
+                    else
+                    {
+                        // Three bytes
+                        payloadSize = (((buffer[0] & 0x3f) << 8) | buffer[1]) + 3;
```
Even if I understood the high bit test on line 1610, I wouldn't understand why you only consider 0x3f bytes in this case, but 0x7f in the previous one.  There's a lot of domain knowledge embedded in these magic numbers.

CIPop replied:
Good questions: unfortunately I don't have the domain knowledge.

bartonjs commented on src/System.Net.Security/src/System/Net/SecureProtocols/_SslStream.cs:
```diff
@@ -0,0 +1,709 @@
+            _InternalBuffer = null;
+        }
+
+        ~_SslStream()
```
As far as I can tell, every SslStream is going to put an _SslStream onto the finalizer queue, why isn't this type Disposable?

CIPop replied:
This is actually new code. `PinnableBufferCache` is an optimization made by the CLR team. @brianrob would have the context.

bartonjs commented on src/System.Net.Security/src/System/Net/SecureProtocols/_SslState.cs:
```diff
@@ -0,0 +1,1745 @@
+                CheckThrow(true);
+                if (_SecureStream == null)
+                {
+                    Interlocked.CompareExchange<_SslStream>(ref _SecureStream, new _SslStream(this), null);
```
_SslStream has a finalizer, so perhaps instead of calling CompareExchange (where multiple threads could call each having instantiated a new _SslStream) it just wants a classic
```C# 
if (_SecureStream == null)
{
    lock (something)
    {
        if (_SecureStream == null)
        {
            _secureStream = new _SslStream(this);
        }
    }
}
```

Ensuring a minimum of finalizer work

CIPop replied:
@bartonjs I'm not willing to risk behavior differences (if not bugs) changing this. 
I'm not exactly sure what parallel calls to Read or Write mean for the public SslStream given the sync implementation.

bartonjs commented on src/System.Net.Security/src/System/Net/SecureProtocols/_SslStream.cs:
```diff
@@ -0,0 +1,709 @@
+            get
+            {
+                return _InternalOffset;
+            }
```
I'm surprised CodeFormatter didn't squeeze these down to
```C#
private int InternalOffset
{
    get { return _InternalOffset; }
}
```

CIPop replied:
I'm not surprised anymore: too much manual work has been required to format code. 
Since this is clear enough I won't do the changes here.

bartonjs commented on src/System.Net.Security/src/System/Net/SecureProtocols/_SslStream.cs:
```diff
@@ -0,0 +1,709 @@
+
+        private void DecrementInternalBufferCount(int decrCount)
+        {
+            _InternalOffset += decrCount;
```
I found it odd that the first line in a method named "Decrement" is a `+=` operation.  Then, having spotted the line I'd expect, this line still seems slightly out of place.  I think I might follow, but some comments wouldn't go amiss.

CIPop replied:
You are correct: it's not very clear from the code. It looks like the `InternalBufferCount` is the remainder of the buffer and `InternalOffset` is a pointer within `InternalBuffer`. When the remainder is 0 the offset is set to the end of the buffer.

Instead of comments (that as we've already seen in this PR go out of sync), I'd prefer UTs to pin down the behavior of this class.

bartonjs commented on src/System.Net.Security/src/System/Net/SecureProtocols/_SslStream.cs:
```diff
@@ -0,0 +1,709 @@
+                int newSize = addSize + curOffset;
+                if (newSize <= PinnableReadBufferSize)
+                {
+                    if (PinnableBufferCacheEventSource.Log.IsEnabled())
```
Is this block reachable?
* AllocateBuffer() doesn't take a size
* and we're only growing in this method
* We already know we need to grow

So my spot-checking tells me that we should be able to assert that `newSize > PinnableReadBufferSize`.  If not, I'm curious as to how we got into that state.

bartonjs replied:
I guess when _InternalBuffer is null?  So maybe that's the assert and comment that I'm wishing I saw.

CIPop replied:
Related to `PinnableBufferCache` @brianrob may be able to explain.
Ideally, UTs should exist to determine the answer and to pin behavior.

bartonjs commented on src/System.Net.Security/src/System/Net/SecureProtocols/_SslStream.cs:
```diff
@@ -0,0 +1,709 @@
+                throw new NotSupportedException(SR.Format(SR.net_io_invalidnestedcall, (asyncRequest != null ? "BeginWrite" : "Write"), "write"));
+            }
+
+            bool failed = false;
```
Personally, I like `failed = true`, then clearing it in the try-block; that way if multiple catches appear they don't all need to set failed to true.

Not a problem here, just food for thought.

CIPop replied:
Won't change but it makes sense.

bartonjs commented on src/System.Net.Security/src/System/Net/SecureProtocols/_SslStream.cs:
```diff
@@ -0,0 +1,709 @@
+
+            // Decrypted data start from "data_offset" offset, the total count can be shrunk after decryption.
+            EnsureInternalBufferSize(0, data_offset + readBytes);
+            DecrementInternalBufferCount(data_offset);
```
Now that I know what both of these do, I have to ask: "Why do we copy the data that we plan to skip?" In particular, if `readBytes` would fit in the pinnable buffer, but `data_offset + readBytes` wouldn't, why heap allocate another array just to skip the first half?

CIPop replied:
Wouldn't touch anything without proper pinning UTs and so Won't fix for now.

bartonjs commented on src/System.Net.Security/src/System/Net/SecureProtocols/_SslStream.cs:
```diff
@@ -0,0 +1,709 @@
+            Buffer.BlockCopy(InternalBuffer, InternalOffset, buffer, offset, readBytes);
+
+            // This will adjust both the remaining internal buffer count and the offset.
+            DecrementInternalBufferCount(readBytes);
```
I think the thing that confuses me is that it's named "BufferCount" instead of "BytesRemaining".  And the method to me feels like it's "Seek{Forward}" or "SkipBytes", not "DecrementCount".

CIPop replied:
My feeling as well (see above comment). It makes sense and clarifies enough that we don't need the comment.

Renamed to SkipBytes().

bartonjs commented on src/System.Net.Security/src/System/Net/_NativeSSPI.cs:
```diff
@@ -0,0 +1,400 @@
+
+        public int EncryptMessage(SafeDeleteContext context, Interop.Secur32.SecurityBufferDescriptor inputOutput, uint sequenceNumber)
+        {
+            int status = (int)Interop.SecurityStatus.InvalidHandle;
```
Why is this declared and initialized out here? It looks like this is never read, it just returns the value from Interop.Secur32.EncryptMessage (so the local could even be eliminated)

CIPop replied:
Ah. I missed this when I did the changes to `DangerousAddRef`. I'll remove and keep `status` within the `try` block. (Will look up all other places.)

bartonjs commented on src/System.Net.Security/src/System/Net/_NativeSSPI.cs:
```diff
@@ -0,0 +1,400 @@
+            context.DangerousAddRef(ref ignore);
+            status = Interop.Secur32.EncryptMessage(ref context._handle, 0, inputOutput, sequenceNumber);
+            context.DangerousRelease();
+
```
In the previous implementation you used a try/finally; here you didn't.

Questions:
1) Can EncryptMessage take a SafeDeleteContext instead of the IntPtr, so the AddRef/Release can be eliminated?
2) Can we reduce the number of identical implementations?  Base class, extension methods, whatever.

CIPop replied:
Thanks for noticing! This is a mistake.

I think the answer to both questions is yes. I believe that will be performed part of the PAL implementation since the interface will change to be platform-agnostic.

bartonjs commented on src/System.Net.Security/src/System/Net/_NativeSSPI.cs:
```diff
@@ -0,0 +1,400 @@
+            bool ignore = false;
+
+            context.DangerousAddRef(ref ignore);
+            const uint SECQOP_WRAP_NO_ENCRYPT = 0x80000001;
```
The const is defined in more than one function, perhaps promote it to a const field?

CIPop replied:
Moved to `Interop.Secur32.SECQOP_WRAP_NO_ENCRYPT`

bartonjs commented on src/System.Net.Security/src/System/Net/_NativeSSPI.cs:
```diff
@@ -0,0 +1,400 @@
+            {
+                bool ignore = false;
+                context.DangerousAddRef(ref ignore);
+                status = Interop.Secur32.DecryptMessage(ref context._handle, inputOutput, sequenceNumber, &qop);
```
Decrypt checked qop != SECQOP_WRAP_NO_ENCRYPT.  Sign used it. Verify just ignores it?

CIPop replied:
We need to add a Unit/Functional test to understand the implications. Tracking with #3114.

bartonjs commented on src/System.Net.Security/src/System/Net/_SSPIWrapper.cs:
```diff
@@ -0,0 +1,873 @@
+{
+    internal static class SSPIWrapper
+    {
+        internal static SecurityPackageInfoClass[] EnumerateSecurityPackages(SSPIInterface SecModule)
```
Naming: Parameter should be camelCased.

CIPop replied:
Fixed throughout the file.

bartonjs commented on src/System.Net.Security/src/System/Net/_SSPIWrapper.cs:
```diff
@@ -0,0 +1,873 @@
+                }
+            }
+
+            if (Logging.On) Logging.PrintInfo(Logging.Web, SR.Format(SR.net_log_sspi_security_package_not_found, packageName));
```
Style: Unbraced ifs should still have the body on the next line, indented.

CIPop replied:
Actually I don't agree with that style: it is error prone and can create the impression that indented lines are actually part of the if block.

It's not in rules within https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/coding-style.md .

I would rather have:
1. If the `if` statement can appear on a single line it can miss braces.
2. Any `if` statement that doesn't fit on a line should _always_ have braces.

bartonjs replied:
Actually, it's rule 1) "... A single line statement block can go without braces but the block <em>must be properly indented on its own line</em> and it must not be nested in other statement blocks that use braces (See issue 381 for examples)." (emphasis mine).

I used to be in the "never have unbraced ifs" camp, now I'm in the "they're only allowed in preconditions" camp (which may be made up of just me).

I'm not necessarily advocating for or against the rule here; just pointing out that we have one.

CIPop replied:
Huh: too obvious so of course I missed it... Thanks for pointing it out.
Will fix all.

bartonjs commented on src/System.Net.Security/src/System/Net/_SSPIWrapper.cs:
```diff
@@ -0,0 +1,873 @@
+            return errorCode;
+        }
+
+        public static int QuerySecurityContextToken(SSPIInterface SecModule, SafeDeleteContext context, out SecurityContextTokenHandle token)
```
About half of these methods use `SecMethod`, the other half `secMethod`.  The latter is definitely preferred.

CIPop replied:
Fixed given the above comment. I would have preferred if the tool did this :disappointed:

bartonjs commented on src/System.Net.Security/src/System/Net/_SSPIWrapper.cs:
```diff
@@ -0,0 +1,873 @@
+        {
+            int errorCode = SecModule.CompleteAuthToken(ref context, inputBuffers);
+
+            if (Logging.On) Logging.PrintInfo(Logging.Web, SR.Format(SR.net_log_operation_returned_something, "CompleteAuthToken()", (Interop.SecurityStatus)errorCode));
```
Lots of non-newlined-unbraced-ifs around here.

CIPop replied:
For all that are not on the same line I've created {} blocks.

CIPop replied:
Changed all, including the ones that were on the same line.

bartonjs commented on src/System.Net.Security/src/System/Net/_SSPIWrapper.cs:
```diff
@@ -0,0 +1,873 @@
+                            errorCode = SecModule.VerifySignature(context, sdcInOut, sequenceNumber);
+                            break;
+
+                        default: throw NotImplemented.ByDesignWithMessage(SR.net_MethodNotImplementedException);
```
Seems like logging the unrecognized op code would be virtuous

CIPop replied:
Added a `GlobalLog.Assert`.

bartonjs commented on src/System.Net.Security/src/System/Net/_SSPIWrapper.cs:
```diff
@@ -0,0 +1,873 @@
+                            iBuffer.token = null;
+                        }
+                        else
+                            checked
```
Please add braces here, when the statements un-indent extra after the closing brace it's a bit confusing.

CIPop replied:
Fixed throughout the files. Please let me know if I missed any in my next commit.

bartonjs commented on src/System.Net.Security/src/System/Net/_SSPIWrapper.cs:
```diff
@@ -0,0 +1,873 @@
+                    }
+
+                    if (errorCode != 0)
+                        if (Logging.On)
```
Can we combine these ifs so that there's not a braced if within an unbraced one?

CIPop replied:
Done.

bartonjs commented on src/System.Net.Security/src/System/Net/_SSPIWrapper.cs:
```diff
@@ -0,0 +1,873 @@
+                    if (errorCode != 0)
+                        if (Logging.On)
+                        {
+                            if (errorCode == 0x90321)
```
Can we name this number?

CIPop replied:
Done.

bartonjs commented on src/System.Net.Security/src/System/Net/_SSPIWrapper.cs:
```diff
@@ -0,0 +1,873 @@
+                    throw new ArgumentException(SR.Format(SR.net_invalid_enum, "ContextAttribute"), "contextAttribute");
+            }
+
+            SafeHandle SspiHandle = null;
```
Naming: Should be sspiHandle

CIPop replied:
Fixed.

bartonjs commented on src/System.Net.Security/src/System/Net/_SSPIWrapper.cs:
```diff
@@ -0,0 +1,873 @@
+                        break;
+
+                    case Interop.Secur32.ContextAttribute.Names:
+                        attribute = Marshal.PtrToStringUni(SspiHandle.DangerousGetHandle());
```
There are a lot of calls to DangerousGetHandle in this switch; should it DangerousAddRef/DangerousRelease before/after?

CIPop replied:
I don't think it's required since `sspiHandle` is defined on the stack. It can't go out of scope as none of these API calls appear to be async.

bartonjs commented on src/System.Net.Security/src/System/Net/_SSPIWrapper.cs:
```diff
@@ -0,0 +1,873 @@
+                        break;
+
+                    case Interop.Secur32.ContextAttribute.LocalCertificate:
+                        goto case Interop.Secur32.ContextAttribute.RemoteCertificate;
```
Why not just do this as a two-label case? The goto feels a little odd given that there's nothing else here and it's "goto next line".

CIPop replied:
Fixed to allow fall-through and added a comment.

bartonjs commented on src/System.Net.Security/src/System/Net/_SSPIWrapper.cs:
```diff
@@ -0,0 +1,873 @@
+
+        internal unsafe SslConnectionInfo(byte[] nativeBuffer)
+        {
+            fixed (void* voidPtr = nativeBuffer)
```
Consider if asserts for nativeBuffer != null and/or nativeBuffer.Length > 28 (Marshal.SizeOf<SslConnectionInfo>()) be useful in tracking down errors

CIPop replied:
Added TODO. The code isn't safe as written: it should cast to a struct instead (`Marshal.PtrToStructure`).

bartonjs commented on src/System.Net.Security/src/System/Net/_SSPIWrapper.cs:
```diff
@@ -0,0 +1,873 @@
+    // TODO (Issue #3114): Move to Interop. 
+    // From Schannel.h
+    [StructLayout(LayoutKind.Sequential)]
+    internal class SslConnectionInfo
```
It's somewhat unusual to see `class` decorated with a StructLayout attribute; did you mean `struct`?

CIPop replied:
Added as part of the TODO comments since converting to `struct` would have implications around allocation and initialization.

bartonjs commented on src/System.Net.Security/src/System/Net/_SSPIWrapper.cs:
```diff
@@ -0,0 +1,873 @@
+                GlobalLog.Print("SecurityPackageInfoClass::.ctor() the pointer is invalid: " + (safeHandle.DangerousGetHandle()).ToString("x"));
+                return;
+            }
+            IntPtr unmanagedAddress = IntPtrHelper.Add(safeHandle.DangerousGetHandle(), SecurityPackageInfo.Size * index);
```
DangerousAddRef/DangerousRelease?

CIPop replied:
Verified callers: the safe handle is created by the caller and kept alive, on the stack for the duration of this ctor. I've added a comment.

bartonjs commented on src/System.Net.Security/src/System/Net/_SecureChannel.cs:
```diff
@@ -0,0 +1,1487 @@
+        internal const int ReadHeaderSize = 5;
+
+        private static volatile X509Store s_MyCertStoreEx;
+        private static volatile X509Store s_MyMachineCertStoreEx;
```
Naming: A couple s_PascalCased statics exist in this type (as opposed to s_camelCase)

CIPop replied:
Fixed.

bartonjs commented on src/System.Net.Security/src/System/Net/_SecureChannel.cs:
```diff
@@ -0,0 +1,1487 @@
+            }
+            finally
+            {
+                if (remoteContext != null)
```
In the try it was also checked for being valid, why not here?

CIPop replied:
You're right: added the check. It doesn't make sense to make an API call with an invalid handle.

bartonjs commented on src/System.Net.Security/src/System/Net/_SecureChannel.cs:
```diff
@@ -0,0 +1,1487 @@
+            {
+                if (remoteContext != null)
+                {
+                    remoteCertificateStore = UnmanagedCertificateContext.GetStore(remoteContext);
```
In the try it was considered a cert, in the finally it's considered a store?

CIPop replied:
Is this duplicate to the one above?

bartonjs replied:
No, I'm confused that in the try block the handle is able to be interpreted as a certificate; but in the finally the same structure is being interpreted as a store.  But it clearly works, so maybe I'm just confused as to what the structure is.

CIPop replied:
This was odd to me too: GetStore is actually returning an `X509Certificate2Collection` in the original code. I've renamed `GetStore` to: `UnmanagedCertificateContext.GetRemoteCertificatesFromStoreContext(remoteContext);`

This is the place where we need to extract certs from `CERT_CONTEXT.hCertStore` but `X509Store(IntPtr)` is missing from CoreFX.

bartonjs commented on src/System.Net.Security/src/System/Net/_SecureChannel.cs:
```diff
@@ -0,0 +1,1487 @@
+                X509Store store = EnsureStoreOpened(_serverMode);
+                if (store != null)
+                {
+                    collectionEx = store.Certificates.Find(X509FindType.FindByThumbprint, certHash, false);
```
FYI, `Find` hasn't been written for Unix yet.

CIPop replied:
Since Unix doesn't have a cert store (afaik), I'm not sure what Find would be like or if it's still something that you need to implement this part.

bartonjs replied:
Stores exist as a concept that we manage; but right now Find always returns no matches.  I can't tell if this code is counting on Find duplicating the certificate handles or not.  (Looping over the collection returns a managed object over the same key handle, calling Find returns a managed object over a different (or at least up-ref'd) key handle).

Since only FindByThumbprint is used I can probably prioritize that to help get things working in a timely manner.

bartonjs commented on src/System.Net.Security/src/System/Net/_SecureChannel.cs:
```diff
@@ -0,0 +1,1487 @@
+                if (store != null)
+                {
+                    collectionEx = store.Certificates.Find(X509FindType.FindByThumbprint, certHash, false);
+                    if (collectionEx.Count > 0 && collectionEx[0].HasPrivateKey)
```
Is the case where the process doesn't have access to the private key in LM\MY, but would for a copy in CU\MY interesting?  `HasPrivateKey` doesn't check permissions.

CIPop replied:
I don't think we should change Desktop behavior without allowing for some control or detection method.

bartonjs commented on src/System.Net.Security/src/System/Net/_SecureChannel.cs:
```diff
@@ -0,0 +1,1487 @@
+                string certHash = null;
+
+                // Protecting from X509Certificate2 derived classes.
+                if (t != typeof(X509Certificate2))
```
This looks duplicate to the `MakeEx` method below, could the first half of the try not just be replaced with
```C#
X509Certificate2 certEx = MakeEx(certificate);
// A nullcheck might be prudent, since MakeEx eats the exceptions and returns null, and this would eat it differently.
string certHash = certEx.Thumbprint;
```

CIPop replied:
Done.

bartonjs commented on src/System.Net.Security/src/System/Net/_SecureChannel.cs:
```diff
@@ -0,0 +1,1487 @@
+
+            if (IsValidContext)
+            {
+                Interop.Secur32.IssuerListInfoEx issuerList = (Interop.Secur32.IssuerListInfoEx)SSPIWrapper.QueryContextAttributes(GlobalSSPI.SSPISecureChannel, _securityContext, Interop.Secur32.ContextAttribute.IssuerListInfoEx);
```
I'm not sure what this is doing... is it equivalent to
```C#
X509Chain chain = new X509Chain();
chain.Build(cert);
return chain.ChainElements.Skip(1).Select(ce => ce.Certificate.SubjectName).ToArray();
```

If so, it would be more portable to write it in managed code... though with whatever validation and online/offline policy you wanted.  And with less System.Linq, probably :smile:.

bartonjs replied:
Okay, looking at how it gets used it seems to be the `CertificateRequest.certificate_authorities` value (http://tools.ietf.org/html/rfc5246#section-7.4.4).  A comment to this effect would be good; and I don't really see a way around doing interop to get that part...

CIPop replied:
Added the RFC comment and renamed this function to `GetRequestCertificateAuthorities` to get closer to the RFC terminology.

bartonjs commented on src/System.Net.Security/src/System/Net/_SecureChannel.cs:
```diff
@@ -0,0 +1,1487 @@
+
+            // Acquire possible Client Certificate information and set it on the handle.
+            X509Certificate clientCertificate = null;   // This is a candidate that can come from the user callback or be guessed when targeting a session restart.
+            ArrayList filteredCerts = new ArrayList();  // This is an intermediate client certs collection that try to use if no selectedCert is available yet.
```
I'm sure a more modern collection type could be used here :smile:.

CIPop replied:
Agreed but optimizations are out of scope for this initial merge. Tracking by #3114

bartonjs commented on src/System.Net.Security/src/System/Net/_SecureChannel.cs:
```diff
@@ -0,0 +1,1487 @@
+                GlobalLog.Print("SecureChannel#" + Logging.HashString(this) + "::AcquireClientCredentials() calling CertificateSelectionCallback");
+
+                X509Certificate2 remoteCert = null;
+                try
```
Why a manual try/finally instead of just a `using`?

CIPop replied:
Won't fix.

bartonjs commented on src/System.Net.Security/src/System/Net/_SecureChannel.cs:
```diff
@@ -0,0 +1,1487 @@
+                        Logging.PrintInfo(Logging.Web, this, SR.Format(SR.net_log_server_issuers_look_for_matching_certs, issuers.Length));
+                }
+
+                for (int i = 0; i < _clientCertificates.Count; ++i)
```
This feels like it could/should be a(t least one) helper method.  Particularly the `i` loop and `ii` loop, with them declared so far apart.
```C#
private static X509Certificate2 FindCandidateByIssuer(X509Certificate2Collection clientCerts, string[] issuers)
{
     ...
}
```
and here:
```C#
if (issuers != null && issuers.Length != 0)
{
    X509Certificate2 candidate = FindCandidateByIssuer(_clientCertificates, issuers);

    if (candidate != null)
    {
        filteredCerts.Add(candidate);
    }
}
else
{
    filteredCerts.AddRange(_clientCertificates);
}
```
It would also allow for much better visualization of the fact that if issuers is null/empty all the client certs are added; otherwise only one cert is added.

bartonjs replied:
And... now I see that I mis-aligned the break when reading the code; so it's really FindCandidate<b>s</b>.  Having that inner loop be a method that returns true/false seems very good to me.

CIPop replied:
Won't fix now. Further refinement can be made in subsequent commits.

bartonjs commented on src/System.Net.Security/src/System/Net/_SecureChannel.cs:
```diff
@@ -0,0 +1,1487 @@
+                            if (certificateEx == null)
+                                continue;
+
+                            GlobalLog.Print("SecureChannel#" + Logging.HashString(this) + "::AcquireClientCredentials() root cert:" + certificateEx.Issuer);
```
I'd expect a check somewhere around here for "if the cert has EKUs, and doesn't have the client EKU, skip it"

bartonjs replied:
A private key is probably also a good thing to check for before deciding that this is the correct cert.  (Maybe both of those things were already done when building _clientCertificates)

CIPop replied:
Maintaining desktop behavior for now. Requires more testing.
Tracking with #3114.

bartonjs commented on src/System.Net.Security/src/System/Net/_SecureChannel.cs:
```diff
@@ -0,0 +1,1487 @@
+            {
+                if (certificate.Handle != IntPtr.Zero)
+                {
+                    certificateEx = new X509Certificate2(certificate.Handle);
```
On Unix the association of the private key to the public certificate is done only in managed code, so the handle ctor will be guaranteed to lose that association.

But, on Unix you shouldn't ever encounter an object whose GetType() == typeof(X509Certificate); so maybe it's fine.

bartonjs commented on src/System.Net.Security/src/System/Net/_SecureChannel.cs:
```diff
@@ -0,0 +1,1487 @@
+            for (int i = 0; i < filteredCerts.Count; ++i)
+            {
+                clientCertificate = filteredCerts[i] as X509Certificate;
+                if ((selectedCert = EnsurePrivateKey(clientCertificate)) != null)
```
Ah, here's the PrivateKey check; so I'll mention the EKU check here, too...

CIPop replied:
Tracking with #3114.

bartonjs commented on src/System.Net.Security/src/System/Net/_SecureChannel.cs:
```diff
@@ -0,0 +1,1487 @@
+                    // So we don't want to reuse **anonymous** cached credential for a new SSL connection if the client has passed some certificate.
+                    // The following block happens if client did specify a certificate but no cached creds were found in the cache.
+                    // Since we don't restart a session the server side can still challenge for a client cert.
+                    if ((object)clientCertificate != (object)selectedCert)
```
If this is trying for {object.}ReferenceEquals, why not just call that method?

CIPop replied:
Won't fix. Similar behavior different syntax.
Changing in one place doesn't make sense. Changing everywhere in the code is out of scope.

bartonjs commented on src/System.Net.Security/src/System/Net/_SecureChannel.cs:
```diff
@@ -0,0 +1,1487 @@
+            GlobalLog.Print("SecureChannel#" + Logging.HashString(this) + "::Encrypt() buffer:");
+            GlobalLog.Dump(buffer, Math.Min(buffer.Length, 128));
+
+            byte[] e_writeBuffer;
```
Naming: Underscores in local variables make for frowning C# developers.

CIPop replied:
Used a regex to search and replace hopefully all remaining instances.

bartonjs commented on src/System.Net.Security/src/System/Net/_SecureChannel.cs:
```diff
@@ -0,0 +1,1487 @@
+            }
+
+            // Encryption using SCHANNEL requires 4 buffers: header, payload, trailer, empty.
+            SecurityBuffer[] securityBuffer = new SecurityBuffer[4];
```
Why not just use an array initializer and avoid all of the numbers?
```C#
SecurityBuffer[] securityBuffer =
{
    new SecurityBuffer(e_writeBuffer, 0, _headerSize, SecurityBufferType.Header),
    /* the other 3 */,
};

CIPop replied:
Won't fix.

bartonjs commented on src/System.Net.Security/src/System/Net/_SecureChannel.cs:
```diff
@@ -0,0 +1,1487 @@
+
+            // Decryption using SCHANNEL requires four buffers.
+            SecurityBuffer[] decspc = new SecurityBuffer[4];
+            decspc[0] = new SecurityBuffer(payload, offset, count, SecurityBufferType.Data);
```
Array initializer

CIPop replied:
Won't fix.

bartonjs commented on src/System.Net.Security/src/System/Net/_SecureChannel.cs:
```diff
@@ -0,0 +1,1487 @@
+            for (int i = 0; i < decspc.Length; i++)
+            {
+                // Successfully decoded data and placed it at the following position in the buffer,
+                if ((errorCode == Interop.SecurityStatus.OK && decspc[i].type == SecurityBufferType.Data)
```
Personally, I'd try to reduce the line length and complexity here by doing this like
```C#
// Useful comment here
SecurityBufferType desiredType =
    (errorCode == Interop.SecurityStatus.OK) ? SecurityBufferType.Data : SecurityBufferType.Extra;

count = 0;

for (int i = 0; i < decspc.Length; i++)
{
    if (decspc[i].type == desiredType)
    {
        // body
    }
}
```

CIPop replied:
Won't fix.

bartonjs commented on src/System.Net.Security/src/System/Net/_SecureChannel.cs:
```diff
@@ -0,0 +1,1487 @@
+                        chain.ChainPolicy.ExtraStore.AddRange(remoteCertificateStore);
+
+                    if (!chain.Build(remoteCertificateEx)       // Build failed on handle or on policy.
+                        && chain.SafeHandle.DangerousGetHandle() == IntPtr.Zero)   // Build failed to generate a valid handle.
```
On Unix `SafeHandle` always returns a wrapped `IntPtr.Zero`.

So when the PAL is injected it'll probably split this condition into the first line of the Windows-specific version.

CIPop replied:
As decided, since this is the certificate handling section, it should move to a separate class which will have common/variations for the various OSs.

bartonjs commented on src/System.Net.Security/src/System/Net/_SecureChannel.cs:
```diff
@@ -0,0 +1,1487 @@
+                        if (remoteCertValidationCallback != null)
+                            Logging.PrintInfo(Logging.Web, this, SR.net_log_remote_cert_user_declared_valid);
+                        else
+                            Logging.PrintInfo(Logging.Web, this, SR.net_log_remote_cert_has_no_errors);
```
Well, it could have had "an error" of RemoteCertificateNotAvailable which can sometimes be considered OK, and would have just printed as "cert has errors".

Didn't bother to look up what would appear in the logs, but the name of the string implies something healthier than the minimums here.

CIPop replied:
This logging was previously public. I'd like to keep things the same to allow for simpler log comparisons between failures when executing against desktop and CoreFX.

bartonjs commented on src/System.Net.Security/src/System/Net/_SecureChannel.cs:
```diff
@@ -0,0 +1,1487 @@
+                rootStore = IntPtr.Zero,
+                phMappers = IntPtr.Zero,
+                palgSupportedAlgs = IntPtr.Zero,
+         
```
Since the message doesn't contain the invalid value, logging it would be good.

CIPop replied:
GitHub places your comment on line 
`+                certContextArray = IntPtr.Zero,`

This is a ctor and so I presume you're referring to the `GlobalLog.Leave` line? There's a TRACE_VERBOSE statement above that.

bartonjs commented on src/System.Net.Security/src/System/Net/_SslSessionsCache.cs:
```diff
@@ -0,0 +1,201 @@
+    // Implements SSL session caching mechanism based on a static table of SSL credentials.
+    internal static class SslSessionsCache
+    {
+        private const int c_CheckExpiredModulo = 32;
```
Style: c_ prefix isn't standard

CIPop replied:
Renaming all to their PascalCase equivalents.

bartonjs commented on src/System.Net.Security/src/System/Net/_SslSessionsCache.cs:
```diff
@@ -0,0 +1,201 @@
+    internal static class SslSessionsCache
+    {
+        private const int c_CheckExpiredModulo = 32;
+        private static Hashtable s_CachedCreds = new Hashtable(32);
```
Can we use something a bit more modern than Hashtable?

CIPop replied:
Tracking with #3114.

bartonjs commented on src/System.Net.Security/src/System/Net/_SslSessionsCache.cs:
```diff
@@ -0,0 +1,201 @@
+        private static Hashtable s_CachedCreds = new Hashtable(32);
+
+        //
+        // Uses cryptographically strong certificate thumbprint comparison.
```
Well... the thumbprint is the SHA-1 hash of the certificate, and that's getting less cryptographically strong as time goes on :smile:.

CIPop replied:
Fixed.

bartonjs commented on src/System.Net.Security/src/System/Net/_SslSessionsCache.cs:
```diff
@@ -0,0 +1,201 @@
+                    if (3 < _CertThumbPrint.Length)
+                        _HashCode ^= (_CertThumbPrint[3] << 24);
+                }
+                _HashCode ^= allowedProtocols;
```
How big are allowedProtocols and encryptionPolicy? Should one of them undergo a `<< 16` to pad out their value in the hashcode?

CIPop replied:
Won't fix. Changing this requires further testing.

bartonjs commented on src/System.Net.Security/src/System/Net/_SslSessionsCache.cs:
```diff
@@ -0,0 +1,201 @@
+            //           the caller of this ctor has to ensure that a user cert object was inspected and
+            //           optionally cloned.
+            //
+            internal SslCredKey(byte[] thumbPrint, int allowedProtocols, EncryptionPolicy encryptionPolicy)
```
While it's Hard(tm) for two legitimate certificates to end up with the same thumbprint, it's not guaranteed unique.  If you think the domain is constricted enough to make you feel safe, that's fine; otherwise some additional material should be used as a tiebreaker.

CIPop replied:
I think you'd have other issues at that time but if you feel this is a real issue please add it to #3114. 
Won't fix: large behavior/feature-level changes are out of scope.

bartonjs commented on src/System.Net.Security/src/System/Net/_SslSessionsCache.cs:
```diff
@@ -0,0 +1,201 @@
+
+            if (cached == null || cached.IsClosed || cached._Target.IsInvalid)
+            {
+                GlobalLog.Print("TryCachedCredential() Not Found, Current Cache Count = " + s_CachedCreds.Count);
```
The logging message here doesn't really match the condition... if it was closed out it was "found", just not "usable"...

CIPop replied:
Fixed. (GlobalLog is not part of previously public logging.)

bartonjs commented on src/System.Net.Security/src/System/Net/_SslSessionsCache.cs:
```diff
@@ -0,0 +1,201 @@
+                        //
+                        //    We won't shrink cache in the case when NO new handles are coming to it.
+                        //
+                        if ((s_CachedCreds.Count % c_CheckExpiredModulo) == 0)
```
Why mod instead of `>=`?  is it not possible to have two add at close enough to the same time to skip right over the modulo?

CIPop replied:
Not sure. Won't fix without adding a UT to ensure the behavior is maintained.

bartonjs commented on src/System.Net.Security/src/System/Net/_SslSessionsCache.cs:
```diff
@@ -0,0 +1,201 @@
+                                if (cached != null)
+                                {
+                                    creds = cached._Target;
+                                    cached.Dispose();
```
Why not just use AddRef/Release on the underlying handle directly and get rid of the middle-man?  The notion of caching here makes it feel like you can end up wanting to share, and I'm not sure how the one-shot wrapper really accomplishes that.

CIPop replied:
Won't fix. We don't have enough tests to ensure the changes would be safe.

bartonjs commented on src/System.Net.Security/src/System/PinnableBufferCache.cs:
```diff
@@ -0,0 +1,552 @@
+                _restockSize = 4;
+            else if (_restockSize < DefaultNumberOfBuffers)
+                _restockSize = DefaultNumberOfBuffers;
+            else if (_restockSize < 256)
```
Const the 4, 256, and 4096

CIPop replied:
Won't fix. The magic numbers are localized within the function that's using them. Tuning cannot be done without changing this function (note the fractional restock growth rate for large restock sizes). 

I believe the code as written is cleaner.

bartonjs commented on src/System.Net.Security/src/System/PinnableBufferCache.cs:
```diff
@@ -0,0 +1,552 @@
+        /// <summary>
+        /// Create a PinnableBufferCache that works on any object (it is intended for OverlappedData)
+        /// </summary>
+        internal PinnableBufferCache(string cacheName, Func<object> factory)
```
Why `Func<object>`? Things will go badly if it doesn't return a `byte[]`, so `Func<byte[]>` seems more apt...

CIPop replied:
Won't fix. Please work with CLR to clarify PinnableBufferCache design questions.

bartonjs commented on src/System.Net.Security/src/System/PinnableBufferCache.cs:
```diff
@@ -0,0 +1,552 @@
+        /// <summary>
+        /// Contains 'good' buffers to reuse.  They are guaranteed to be Gen 2 ENFORCED!
+        /// </summary>
+        private ConcurrentStack<object> _freeList = new ConcurrentStack<object>();
```
Why are all the generics over object?

CIPop replied:
Won't fix. Please work with CLR to clarify PinnableBufferCache design questions.

bartonjs commented on src/System.Net.Security/src/System/PinnableBufferCache.cs:
```diff
@@ -0,0 +1,552 @@
+        /// To implement trimming we sometimes put aged buffers in here as a place to 'park' them
+        /// before true deletion.  
+        /// </summary>
+        private List<object> _notGen2;
```
I don't understand why there's a preference for handing back things in gen2 over gen1 (or gen0).  Really there's a lot about this type that feels like it's trying to be clever in ways that it doesn't explain, which makes me wonder if it's engineered to last through the ages, or was just trying to match some performance goal from long ago and the universe has moved on making this now inefficient.

This feels like it should just be a ConcurrentQueue (or stack) and when empty it adds 10 (or whatever) and on a Gen2 collect it purges part of the cache if it's over some size; but no caring about what generation any particular object is in.

CIPop replied:
Won't fix. Please work with CLR to clarify PinnableBufferCache design questions.

bartonjs commented on src/System.Net.Security/src/System/Security/Authentication/ExtendedProtection/ServiceNameCollection.cs:
```diff
@@ -0,0 +1,205 @@
+
+        public ServiceNameCollection Merge(string serviceName)
+        {
+            ArrayList newServiceNames = new ArrayList();
```
Can these methods be upgraded to something newer than ArrayList?

In particular, the `Merge(IEnumerable)` seems like it could benefit from a HashSet, unless the order is guaranteed (even then, a HashSet for duplicate detection and a List for ordering seems +EV)

CIPop replied:
Tracked by #3114.

bartonjs commented on src/System.Net.Security/src/System/Security/Authentication/ExtendedProtection/ServiceNameCollection.cs:
```diff
@@ -0,0 +1,205 @@
+            }
+
+            // Separate out the prefix
+            int shashIndex = inputServiceName.IndexOf('/');
```
Naming (typo): slashIndex (vs shashIndex)

CIPop replied:
Corrected.

bartonjs commented on src/System.Net.Security/src/System/Security/Authentication/ExtendedProtection/ServiceNameCollection.cs:
```diff
@@ -0,0 +1,205 @@
+                }
+
+                // Check for port.
+                int colonIndex = hostAndPort.LastIndexOf(':'); // Allow IPv6 addresses.
```
Is the comment (// Allow IPv6 addresses) a "this works", or a TODO? It feels like the latter, is there a tracking issue?

bartonjs replied:
Presumably it would be of the form [addr]:port (literal square braces), like in URIs? http://tools.ietf.org/html/rfc3986#section-3.2.2

CIPop replied:
It's not a TODO, but an explanation on why that line exists.
This is not a URI, but an [SPN](https://msdn.microsoft.com/en-us/library/ms677601(v=vs.85).aspx) so I can't be 100% sure.

CIPop commented:
@bartonjs Submitted an updated version that hopefully addresses most of your comments.
Thanks _a lot_ for your time and all the feedback provided!

Tracking #3114 is linked to this PR to fix remaining complex changes that would require either Unit or Functional tests to ensure behavior pinning.

bartonjs commented:
@CIPop The comments streaming back via email seemed reasonable.  I'll make a passive comment that either `SslStream` or `_SslStream` (I can't remember) still has underscore-pascal-cased fields; but that sort of stuff can just be cleaned up later.  This PR already makes the browser sad enough to load :).

Thanks for moving some of the comments into the tracking issue; I would've been fine with all of my comments getting that sort of response :).

LGTM.  (If CI fails due to #3237 again, just edit the original description and remove the blank line and @-references after)

CIPop commented:
Thanks @stephentoub. I'll remove the last commit and rebase after your fix is merged.

stephentoub commented:
> I'll remove the last commit and rebase after your fix is merged.

Thanks. It's now been merged.

CIPop commented:
Thanks @stephentoub! I've removed the last commit and will merge the PR if everything checks out.

jasonwilliams200OK commented:
:bulb:  Couple of merge commits cropped up by the recent PR merges.

CIPop commented:
Thanks for the heads up @jasonwilliams200OK!
